src/network/networkd-network.c

```diff
@@ -36,6 +36,65 @@
 #include "string-util.h"
 #include "util.h"
 
+static void network_config_hash_func(const void *p, struct siphash *state) {
+        const NetworkConfigSection *c = p;
+        unsigned filename_size = strlen(c->filename);
+
+        siphash24_compress(c->filename, filename_size, state);
+        siphash24_compress(&filename_size, sizeof(filename_size), state);
+        siphash24_compress(&c->line, sizeof(c->line), state);
+}
+
+static int network_config_compare_func(const void *a, const void *b) {
+        const NetworkConfigSection *x = a, *y = b;
+        int r;
+
+        r = memcmp(x->filename, y->filename, MIN(strlen(x->filename), strlen(y->filename)));
+        if (r != 0)
+                return r;
+
+        if (strlen(x->filename) < strlen(y->filename))
+                return -1;
+
+        if (strlen(x->filename) > strlen(y->filename))
+                return 1;
+
+        if (x->line < y->line)
+                return -1;
+
+        if (x->line > y->line)
+                return 1;
+
+        return 0;
+}
+
+const struct hash_ops network_config_hash_ops = {
+        .hash = network_config_hash_func,
+        .compare = network_config_compare_func,
+};
+
+int network_config_section_new(const char *filename, unsigned line, NetworkConfigSection **s) {
+        _cleanup_network_config_section_free_  NetworkConfigSection *cs = NULL;
+
+        cs = malloc0(offsetof(NetworkConfigSection, filename) + strlen(filename) + 1);
+        if (!cs)
+                return -ENOMEM;
+
+        cs->line = line;
```

Comment:
What about `filename`? It needs to be copied into `cs` too.

You can drop `_cleanup_` for `cs`, once the allocation succeeds, function cannot fail.